pinot-broker/src/main/java/org/apache/pinot/broker/routing/adaptiveserverselector/PriorityGroupInstanceSelector.java

```diff
@@ -0,0 +1,229 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.broker.routing.adaptiveserverselector;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.pinot.broker.routing.instanceselector.SegmentInstanceCandidate;
+
+
+/**
+ * A server selector that implements priority-based server selection based on replica groups.
+ * This selector works in conjunction with an {@link AdaptiveServerSelector} to provide
+ * a two-level selection strategy:
+ * <ol>
+ *   <li>First level: Select servers based on replica group priorities</li>
+ *   <li>Second level: Use adaptive selection within the chosen replica group</li>
+ * </ol>
+ *
+ * <p>The selector maintains the following invariants:</p>
+ * <ul>
+ *   <li>Servers from preferred replica groups are always selected before non-preferred groups</li>
+ *   <li>Within each replica group, servers are selected using adaptive selection criteria</li>
+ *   <li>When no preferred groups are available, falls back to non-preferred groups</li>
+ * </ul>
+ */
+public class PriorityGroupInstanceSelector {
+
+  /** The underlying adaptive server selector used for selection within replica groups */
+  private final AdaptiveServerSelector _adaptiveServerSelector;
```

Comment:
We are tying this feature to adaptive server selector. But for certain use-cases we may have to disable it (e.g. for colocated joins).

Reply:
(1) we need follow ups to support this query options for non adaptive server selector.
(2) does the colocated join expect fixedReplicas?